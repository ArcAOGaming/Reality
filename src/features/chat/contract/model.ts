import { ArweaveId } from "@/features/arweave/lib/model";
import { z } from "zod";

// Placeholder
// TODO: Define this properly
export const Message = z.object({
  MessageId: ArweaveId,
  Timestamp: z.number(),
  AuthorId: ArweaveId,
  AuthorName: z.string(),
  Content: z.string(),
});
export type Message = z.infer<typeof Message>;

export const MessageCreate = Message.omit({
  MessageId: true,
  Timestamp: true,
  AuthorId: true,
});
export type MessageCreate = z.infer<typeof MessageCreate>;
