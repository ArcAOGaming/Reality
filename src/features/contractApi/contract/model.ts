// Placeholder
// TODO: Define this properly
import { RJSFSchema } from '@rjsf/utils';
import { z } from 'zod';

export const ApiSchemaMethod = z.object({
  Title: z.string(),
  Description: z.string(),
  Schema: z.object({
    Tags: z.custom<RJSFSchema>(),
  }),
});
export type ApiSchemaMethod = z.infer<typeof ApiSchemaMethod>;

export const ApiSchema = z.record(ApiSchemaMethod);
export type ApiSchema = z.infer<typeof ApiSchema>;
