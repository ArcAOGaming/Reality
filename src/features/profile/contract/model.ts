import { ArweaveId } from "@/features/arweave/lib/model";
import { z } from "zod";

export const ProfileInfo = z.object({
  ProfileId: ArweaveId,
  Username: z.string(),
  ProfileImage: z.string(),
  CoverImage: z.string(),
  Description: z.string(),
  DisplayName: z.string(),
});
export type ProfileInfo = z.infer<typeof ProfileInfo>;

export const ProfileInfoCreate = ProfileInfo.omit({
  ProfileId: true,
  Username: true,
})
  .extend({
    UserName: z.string(),
    DateCreated: z.number(),
    DateUpdated: z.number(),
  })
  .partial({
    // Username: true,
    // ProfileImage: true,
    // CoverImage: true,
    // Description: true,
    // DisplayName: true,
  });
export type ProfileInfoCreate = z.infer<typeof ProfileInfoCreate>;

export const ProfileInfoUpdate = ProfileInfoCreate.omit({
  DateCreated: true,
}).partial();
export type ProfileInfoUpdate = z.infer<typeof ProfileInfoUpdate>;
