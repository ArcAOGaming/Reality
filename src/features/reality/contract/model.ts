import { z } from "zod";
import { _2dTileParams } from "./_2dTile";
import { ArweaveAddress } from "@/features/arweave/lib/model";

export const RealityVector = z.array(z.number());
export type RealityVector = z.infer<typeof RealityVector>;

export const RealityInfo = z.object({
  Parent: z.string(),
  Name: z.string(),
  Dimensions: z.number(),
  "Render-With": z.string(),
});
export type RealityInfo = z.infer<typeof RealityInfo>;

export const RealityParameterBounds = z.object({
  Lower: z.array(z.number()),
  Upper: z.array(z.number()),
});

export const RealityParameters = z.object({
  Bounds: z.optional(RealityParameterBounds),
  "2D-Tile-0": z.optional(_2dTileParams),
});
export type RealityParameters = z.infer<typeof RealityParameters>;

export const RealityEntityType = z.enum(["Unknown", "Avatar", "Hidden"]);
export type RealityEntityType = z.infer<typeof RealityEntityType>;

export const RealityEntityPosition = RealityVector;
export type RealityEntityPosition = z.infer<typeof RealityEntityPosition>;

export const DefaultInteraction = z.object({
  Type: z.literal("Default"),
});
export type DefaultInteraction = z.infer<typeof DefaultInteraction>;

export const Warp = z.object({
  Type: z.literal("Warp"),
  Size: z.optional(z.array(z.number())),
});
export type Warp = z.infer<typeof Warp>;

export const SchemaForm = z.object({
  Type: z.literal("SchemaForm"),
  Id: z.string(),
});
export type SchemaForm = z.infer<typeof SchemaForm>;

export const SchemaExternalForm = z.object({
  Type: z.literal("SchemaExternalForm"),
  Id: z.string(),
});
export type SchemaExternalForm = z.infer<typeof SchemaForm>;

export const RealityEntityMetadata = z.intersection(
  z.object({
    Interaction: z.optional(
      z.discriminatedUnion("Type", [
        DefaultInteraction,
        Warp,
        SchemaForm,
        SchemaExternalForm,
      ]),
    ),
    ProfileId: z.optional(z.string()),
    DisplayName: z.optional(z.string()),
    SkinNumber: z.optional(z.number()),
    SpriteTxId: z.optional(z.string()),
  }),
  z.record(z.string(), z.any()),
);
export type RealityEntityMetadata = z.infer<typeof RealityEntityMetadata>;

export const RealityEntity = z.object({
  Position: RealityEntityPosition,
  Type: RealityEntityType,
  Metadata: z.optional(RealityEntityMetadata),
  StateCode: z.optional(z.number()),
});
export type RealityEntity = z.infer<typeof RealityEntity>;

export const RealityEntityCreate = z.object({
  Type: z.optional(RealityEntityType),
  Position: z.optional(RealityEntityPosition),
  Metadata: z.optional(RealityEntityMetadata),
});
export type RealityEntityCreate = z.infer<typeof RealityEntityCreate>;

export const RealityEntities = z.record(ArweaveAddress, RealityEntity);
export type RealityEntities = z.infer<typeof RealityEntities>;

export const RealityEntityKeyed = RealityEntities.refine(
  (entities) => Object.keys(entities).length === 1,
  {
    message: "Expected exactly one entity",
  },
);
export type RealityEntityKeyed = z.infer<typeof RealityEntityKeyed>;
