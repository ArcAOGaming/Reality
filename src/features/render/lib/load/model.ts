import { ProfileRegistryClient } from "@/features/profile/contract/profileRegistryClient";
import { RealityClient } from "@/features/reality/contract/realityClient";

export type WorldState = {
  info: Awaited<ReturnType<RealityClient["readInfo"]>>;
  parameters: Awaited<ReturnType<RealityClient["readParameters"]>>;
  entities: Awaited<ReturnType<RealityClient["readEntitiesStatic"]>>;
  profiles: Awaited<ReturnType<ProfileRegistryClient["readProfiles"]>>;
};
