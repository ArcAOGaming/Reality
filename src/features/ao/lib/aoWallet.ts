import { createDataItemSigner } from "@permaweb/aoconnect";

export type AoSigner = ReturnType<typeof createDataItemSigner>

export type AoWallet = {
  address: string;
  signer: AoSigner;
}

export type AoWalletConnectionResult = {
  success: true;
  result: AoWallet;
} | {
  success: false;
  error: unknown;
}

// TODO: Import these from somewhere more official
export type PermissionType =
  | "ACCESS_ADDRESS"
  | "ACCESS_PUBLIC_KEY"
  | "ACCESS_ALL_ADDRESSES"
  | "SIGN_TRANSACTION"
  | "ENCRYPT"
  | "DECRYPT"
  | "SIGNATURE"
  | "ACCESS_ARWEAVE_CONFIG"
  | "DISPATCH";

export interface AppInfo {
  name?: string;
  logo?: string;
}

export interface GatewayConfig {
  host: string;
  port: number;
  protocol: "http" | "https";
}

export type AoWalletConnector = (
  permissions: PermissionType[],
  appInfo?: AppInfo,
  gateway?: GatewayConfig,
) => Promise<AoWalletConnectionResult>;
