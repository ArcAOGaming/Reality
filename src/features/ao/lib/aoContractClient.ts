import { z } from "zod";
import { AoClient, Message } from "./aoClient";
import { AoWallet } from "./aoWallet";
import { connect } from "@permaweb/aoconnect";

class AoContractError extends Error {
  error: unknown | undefined;

  constructor(message: string, error?: unknown) {
    super(message);
    this.name = "AoContractError";
    this.error = error;
  }
}

export type ReadArgs = Pick<Parameters<AoClient["dryrun"]>[0], "tags" | "data" | "anchor">

export type SendArgs = Pick<Parameters<AoClient["message"]>[0], "tags" | "data" | "anchor">

export type AoContractClient = {
  processId: string;
  aoClient: AoClient;
  aoWallet: AoWallet;

  dryrunReadReplyOptional: (readArgs: ReadArgs) => Promise<Message | undefined>;
  dryrunReadReplyOne: (readArgs: ReadArgs) => Promise<Message>;
  dryrunReadReplyOneJson: <T>(readArgs: ReadArgs, schema?: z.Schema) => Promise<T>;
  message: (sendArgs: SendArgs) => Promise<string>;
}

export const createAoContractClient = (
  processId: string,
  aoClient: AoClient,
  aoWallet: AoWallet,
): AoContractClient => {
  const dryrunReadReplyOptional = async (readArgs: ReadArgs) => {
    const result = await aoClient.dryrun({
      ...readArgs,
      process: processId,
      Target: processId,
      Owner: aoWallet.address,
    });
    const messages = result.Messages as Array<Message>;

    if (messages.length === 0) {
      return undefined;
    }

    const reply = messages.find((msg) => msg.Target === aoWallet.address);
    return reply;
  }

  const dryrunReadReplyOne = async (readArgs: ReadArgs) => {
    const reply = await dryrunReadReplyOptional(readArgs);
    if (!reply) {
      throw new AoContractError("No reply");
    }
    return reply;
  }

  const dryrunReadReplyOneJson = async (readArgs: ReadArgs, schema?: z.Schema) => {
    const reply = await dryrunReadReplyOne(readArgs);
    
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let json: any;
    try {
      json = JSON.parse(reply.Data);
    } catch (error) {
      throw new AoContractError("Invalid JSON", error);
    }
    if (schema) {
      const result = schema.safeParse(json);
      if (!result.success) {
        throw new AoContractError("JSON does not match schema", result.error);
      }
      return result.data;
    }
    return json;
  }

  const message = async (sendArgs: SendArgs) => 
    aoClient.message({
      ...sendArgs,
      process: processId,
      signer: aoWallet.signer,
    });

  return {
    processId,
    aoClient,
    aoWallet,

    dryrunReadReplyOptional,
    dryrunReadReplyOne,
    dryrunReadReplyOneJson,
    message,
  }
}

export const createAoContractClientForProcess = (wallet: AoWallet) => (processId: string) => {
  const aoClient = connect();
  return createAoContractClient(processId, aoClient, wallet);
}
export type AoContractClientForProcess = ReturnType<typeof createAoContractClientForProcess>;
