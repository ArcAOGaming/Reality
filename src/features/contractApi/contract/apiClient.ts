import { AoContractClient, createAoContractClient } from "../../ao/lib/aoContractClient";
import { ApiSchema } from "./model";
import { AoWallet } from "@/features/ao/lib/aoWallet";
import { connect } from "@permaweb/aoconnect";

export type ApiClient = {
  aoContractClient: AoContractClient;

  // Reads
  readApi(): Promise<ApiSchema>;
}

// Placeholder
// TODO: Define these methods properly
export const createApiClient = (
  aoContractClient: AoContractClient,
): ApiClient => ({
  aoContractClient: aoContractClient,

  // Read
  readApi: () => aoContractClient.dryrunReadReplyOneJson<ApiSchema>({
    tags: [{ name: "Read", value: "Api" }],
  }, /* ApiSchema */),
});

export const createApiClientForProcess = (wallet: AoWallet) => (processId: string) => {
  const aoContractClient = createAoContractClient(processId, connect(), wallet);
  return createApiClient(aoContractClient);
}
