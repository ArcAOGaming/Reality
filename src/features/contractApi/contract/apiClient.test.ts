import { describe, test, expect, beforeAll } from "vitest";
import { createApiClient } from "./apiClient";
import { AoContractClient, createAoContractClient } from "@/features/ao/lib/aoContractClient";
import { loadTestWallet } from "@/features/ao/test/lib/fsWallet";
import { AoWallet } from "@/features/ao/lib/aoWallet";
import { connect } from "@permaweb/aoconnect";
import { ApiSchema } from "./model";

describe('createApiClient', () => {
  let testWallet: AoWallet;
  let apiTestAoContractClient: AoContractClient;

  beforeAll(async () => {
    testWallet = await loadTestWallet();
    apiTestAoContractClient = createAoContractClient(import.meta.env.VITE_LLAMAASSISTANT_PROCESS_ID, connect(), testWallet);
  })

  test('creates client', async () => {
    const client = createApiClient(apiTestAoContractClient)
    expect(client).toMatchSnapshot();
  })

  test('read Api', async () => {
    const client = createApiClient(apiTestAoContractClient)

    const api = await client.readApi();
    expect(ApiSchema.safeParse(api).success).toBe(true);
    expect(api).toMatchSnapshot();
  })
})
