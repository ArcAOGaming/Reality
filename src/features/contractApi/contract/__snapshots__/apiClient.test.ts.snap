// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createApiClient > creates client 1`] = `
{
  "aoContractClient": {
    "aoClient": {
      "assign": [Function],
      "dryrun": [Function],
      "message": [Function],
      "monitor": [Function],
      "result": [Function],
      "results": [Function],
      "spawn": [Function],
      "unmonitor": [Function],
    },
    "aoWallet": {
      "address": "HHz5nNVEzkALjmZq8cdIdubMBJCpIFOvzO5uHfU5nQI",
      "anonymous": false,
      "signer": [Function],
      "type": "Keyfile",
    },
    "dryrunReadReplyOne": [Function],
    "dryrunReadReplyOneJson": [Function],
    "dryrunReadReplyOptional": [Function],
    "message": [Function],
    "processId": "y_obnQk5mkphKlulM7v1Xyn6IhJKZGhP_BC1qLJq46w",
  },
  "readApi": [Function],
}
`;

exports[`createApiClient > read Api 1`] = `
{
  "Petition": {
    "Description": "You must stake some $CRED for a chance to earn $LLAMA",
    "Schema": {
      "Tags": {
        "properties": {
          "Action": {
            "const": "Petition",
            "type": "string",
          },
          "Offering": {
            "maximum": 100,
            "minimum": 1,
            "type": "integer",
          },
          "Prompt": {
            "maxLength": 100,
            "minLength": 2,
            "type": "string",
          },
        },
        "required": [
          "Action",
          "Offering",
          "Prompt",
        ],
        "type": "object",
      },
    },
    "Title": "Petition the LlamaFed",
  },
}
`;
